logbus/logstreamer/logstreamer_orchestrator.go

```diff
@@ -0,0 +1,170 @@
+package logstreamer
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/earthly/cloud-api/logstream"
+	"github.com/hashicorp/go-multierror"
+)
+
+type LogstreamOrchestrator struct {
+	buildID         string
+	bus             LogBus
+	c               CloudClient
+	initialManifest *logstream.RunManifest
+
+	doneMU sync.Mutex
+	mu     sync.Mutex
+	doneCH chan struct{}
+
+	errMu       sync.Mutex
+	errors      []error
+	started     atomic.Bool
+	retries     int
+	streamer    *LogStreamer
+	deltas      *deltasIter
+	verbose     bool
+	deltaBuffer int
+}
+
+type LOpt func(*LogstreamOrchestrator) *LogstreamOrchestrator
+
+// WithVerbose sets the verbose option on the LogstreamOrchestrator
+func WithVerbose(verbose bool) func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+	return func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+		orchestrator.verbose = verbose
+		return orchestrator
+	}
+}
+
+func WithDeltaBuffer(buffer int) func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+	return func(orchestrator *LogstreamOrchestrator) *LogstreamOrchestrator {
+		orchestrator.deltaBuffer = buffer
+		return orchestrator
+	}
+}
+
+func NewLogstreamOrchestrator(bus LogBus, c CloudClient, initialManifest *logstream.RunManifest, opts ...LOpt) *LogstreamOrchestrator {
+	ls := &LogstreamOrchestrator{
+		buildID:         initialManifest.GetBuildId(),
+		bus:             bus,
+		c:               c,
+		initialManifest: initialManifest,
+		retries:         10,
+		deltaBuffer:     DefaultBufferSize,
+	}
+	for _, o := range opts {
+		ls = o(ls)
+	}
+	return ls
+}
+
+// StartLogstreamer will start streaming to the cloud retrying up the retry count
+// Callers should listen to Done to be notified when the streaming contract completes
+// StartLogstreamer may only be called once
+func (l *LogstreamOrchestrator) StartLogstreamer(ctx context.Context) {
+	if l.started.Swap(true) {
+		// Can only start once
+		return
+	}
+	go func() {
+		for i := 0; i < l.retries; i++ {
+			l.start()
+			l.CloseLastLogstreamer()
+
+			l.mu.Lock()
+			l.deltas = newDeltasIter(l.deltaBuffer, l.initialManifest, l.verbose)
+			l.streamer = New(l.c, l.buildID, l.deltas)
+			l.mu.Unlock()
```

Comment:
It always makes me nervous when we can't `defer` an `Unlock()`. Can we move this to a separate method, e.g. `l.reset()` so that we can `defer l.mu.Unlock()` in that function?